### ISSUE TITLE
Match local networks when determining if a client is local

### ISSUE THREAD

**Issue by brianjmurrell** [Sonarr/Sonarr#6534]
### Is there an existing issue for this?

- [X] I have searched the existing open and closed issues

### Is your feature request related to a problem? Please describe

When I connect to my Sonarr instance on my LAN using it's Internet-routable (IPv6, but would apply equally to IPv4 -- I just don't have an Internet routable IPv4 network assigned to me) address I am required to authenticate even though I have opted to not require authentication for _local addresses_.

### Describe the solution you'd like

https://github.com/Sonarr/Sonarr/blob/main/src/NzbDrone.Common/Extensions/IpAddressExtensions.cs should be expanded to match the client address against all address/netmask pairs assigned to interfaces on the Sonarr machine for a match.

In fact once that is done, the matching of unroutable and link-local networks can be eliminated because any of those that apply will be assigned to an interface on the machine and the above proposed matching will apply to them automagically.

### Describe alternatives you've considered

There are none.

**Comment by markus101** [comment]
If you're connecting via an internet routable IPv6 address then Sonarr's detection is correct as it's not a _local_ IP address. If Sonarr included all IP addresses assigned to network interfaces on the machine it's running on then it'd effectively require no authentication for requests made to a public IP address if it was directly on the internet, like a seedbox, VPS or dedicated machine.

**Comment by brianjmurrell** [comment]
> If you're connecting via an internet routable IPv6 address then Sonarr's detection is correct as it's not a _local_ IP address.

I would beg to differ.  You seem to be conflating two different concepts:
- routable/non-routable addressing
- local network determination

They are not the same thing.  The former can be used as a proxy for the latter but even that is not really precise.

The definition of a _local address_ is an address that is in the same logical network.  I.e. determining if the address of the client is within any of the ip/netmask pairs assigned on the server regardless of whether they are routable or not because those addresses **have** to be on the same subnet (i.e. local LAN) to even work.

> If Sonarr included all IP addresses assigned to network interfaces on the machine it's running on then it'd effectively require no authentication for requests made to a public IP address if it was directly on the internet, like a seedbox, VPS or dedicated machine.

This is completely un-true.  Even if the Sonarr machine were on a VPS with a publicly routable IP address, any client that is not on the same subnet (i.e. that were somewhere else on the Internet) would not match any of the ip-address/netmask pairs on the Sonarr machine and would therefore be subject to authentication.

For example, a Sonarr machine on my LAN with an IP address of 2001:123:4567:0:92df:5ff3:615f:6019/64 and a client with address of 2001:123:4567:0:90cc:a629:cf42:5877 are on the same network and considered "local".  The reason is that 2001:123:4567:0:90cc:a629:cf42:5877 with the lower 64 bits masked out == 2001:123:4567::/64 which is the same for 2001:123:4567:0:92df:5ff3:615f:6019/64.  Only machines on the same LAN can and will match this.  This is also true of routable IPv4 addresses such as 198.1.2.1/24 and 198.1.2.2/24.  They key is in matching the client and server addresses with their subnet bits masked out (I.e. the lower 8 bits for a /24 IPv4 network).

Any other host on the Internet not in my subnet would not match as they would not be allowed to have an address in the 2001:123:4567::/64 network and it wouldn't work even if they tried.  The Internet would not route traffic back to them.

Here's a C# method to determine if an (i.e. client) IP address is within a subnet of (i.e. a, say, Sonarr) server (complete with a stackoverflow reference):
https://www.ryadel.com/en/check-ip-address-within-given-subnet-mask-c-sharp/

Here's a SO Q/A for finding the IP address/networks on a machine:
https://stackoverflow.com/questions/5271724/get-all-ip-addresses-on-machine

The [top answer](https://stackoverflow.com/a/5272099) is the right one.  The other(s) that use DNS to determine a machines IP addresses are wrong (or at least sub-optimal, non-canonical, etc.).

Using these methods provides a canonical answer to "is the client local?" and avoids using (not always correct) proxies for that question such "does the client use a non-routable address?".

**Comment by brianjmurrell** [comment]
Here's a entire example program.  Please ignore my ignorance of C#.  This is in fact my very first C# program.  :-)

```c#
using System;
using System.Net.NetworkInformation;
using System.Net;
using System.Net.Sockets;
using System.Collections;

class Program
{
    /// <summary>
    /// Returns TRUE if the given IP address is contained in the given subnetmask, FALSE otherwise.
    /// Examples:
    /// - IsInSubnet("192.168.5.1", "192.168.5.85/24") -> TRUE
    /// - IsInSubnet("192.168.5.1", "192.168.5.85/32") -> FALSE
    /// ref.: https://stackoverflow.com/a/56461160
    /// </summary>
    /// <param name="address">The IP Address to check</param>
    /// <param name="subnetMask">The SubnetMask</param>
    /// <returns></returns>
    /// <exception cref="NotSupportedException"></exception>
    /// <exception cref="ArgumentException"></exception>
    public bool IsInSubnetMask(string ipAddress, string subnetMask)
    {
        var address = IPAddress.Parse(ipAddress);
        var slashIdx = subnetMask.IndexOf("/");
        if (slashIdx == -1)
            // We only handle netmasks in format "IP/PrefixLength".
            throw new NotSupportedException("Only SubNetMasks with a given prefix length are supported.");

        // First parse the address of the netmask before the prefix length.
        var maskAddress = IPAddress.Parse(subnetMask.Substring(0, slashIdx));

        if (maskAddress.AddressFamily != address.AddressFamily)
            // We got something like an IPV4-Address for an IPv6-Mask. This is not valid.
            return false;

        // Now find out how long the prefix is.
        int maskLength = int.Parse(subnetMask.Substring(slashIdx + 1));

        if (maskLength == 0)
            return true;

        if (maskLength < 0)
            throw new NotSupportedException("A Subnetmask should not be less than 0.");

        if (maskAddress.AddressFamily == AddressFamily.InterNetwork)
        {
            var maskAddressBits = BitConverter.ToUInt32(maskAddress.GetAddressBytes().Reverse().ToArray(), 0);
            var ipAddressBits = BitConverter.ToUInt32(address.GetAddressBytes().Reverse().ToArray(), 0);
            uint mask = uint.MaxValue << (32 - maskLength);

            // https://stackoverflow.com/a/1499284/3085985
            // Bitwise AND mask and MaskAddress, this should be the same as mask and IpAddress
            // as the end of the mask is 0000 which leads to both addresses to end with 0000
            // and to start with the prefix.
            return (maskAddressBits & mask) == (ipAddressBits & mask);
        }

        if (maskAddress.AddressFamily == AddressFamily.InterNetworkV6)
        {
            // Convert the mask address to a BitArray. Reverse the BitArray to compare the bits of each byte in the right order.
            var maskAddressBits = new BitArray(maskAddress.GetAddressBytes().Reverse().ToArray());

            // And convert the IpAddress to a BitArray. Reverse the BitArray to compare the bits of each byte in the right order.
            var ipAddressBits = new BitArray(address.GetAddressBytes().Reverse().ToArray());
            var ipAddressLength = ipAddressBits.Length;

            if (maskAddressBits.Length != ipAddressBits.Length)
                throw new ArgumentException("Length of IP Address and Subnet Mask do not match.");

            // Compare the prefix bits.
            for (var i = ipAddressLength - 1; i >= ipAddressLength - maskLength; i--)
                if (ipAddressBits[i] != maskAddressBits[i])
                    return false;

            return true;
        }

        return false;
    }

    public static void Main(string[] args)  
    {  
        Program pr = new Program(); // Creating a class Object  

        foreach (NetworkInterface netInterface in 
            NetworkInterface.GetAllNetworkInterfaces())
        {
            Console.WriteLine("Name: " + netInterface.Name);
            Console.WriteLine("Description: " + netInterface.Description);
            Console.WriteLine("Addresses: ");
        
            IPInterfaceProperties ipProps = netInterface.GetIPProperties();
        
            foreach (UnicastIPAddressInformation addr in ipProps.UnicastAddresses)
            {
                Console.WriteLine(" " + addr.Address.ToString() + "/" +  addr.PrefixLength);
                Console.WriteLine("   Is 10.75.22.10 local? " + pr.IsInSubnetMask("10.75.22.10", addr.Address.ToString() + "/" +  addr.PrefixLength));
                Console.WriteLine("   Is 2001:123:4567:0:92df:5ff3:615f:6019 local? " + pr.IsInSubnetMask("2001:123:4567:0:92df:5ff3:615f:6019", addr.Address.ToString() + "/" +  addr.PrefixLength));
            }
        
            Console.WriteLine("");
        }
    }
}
```

On my local machine, here is an example of it running:

```
Name: lo
Description: lo
Addresses: 
 127.0.0.1/8
   Is 10.75.22.10 local? False
   Is 2001:123:4567:0:92df:5ff3:615f:6019 local? False
 ::1/128
   Is 10.75.22.10 local? False
   Is 2001:123:4567:0:92df:5ff3:615f:6019 local? False

Name: enp2s0
Description: enp2s0
Addresses: 

Name: virbr1
Description: virbr1
Addresses: 
 192.168.100.1/24
   Is 10.75.22.10 local? False
   Is 2001:123:4567:0:92df:5ff3:615f:6019 local? False
 fd31:aeb1:48de::1/64
   Is 10.75.22.10 local? False
   Is 2001:123:4567:0:92df:5ff3:615f:6019 local? False

Name: virbr0
Description: virbr0
Addresses: 
 192.168.122.1/24
   Is 10.75.22.10 local? False
   Is 2001:123:4567:0:92df:5ff3:615f:6019 local? False

Name: pc_bridge
Description: pc_bridge
Addresses: 
 10.75.22.1/24
   Is 10.75.22.10 local? True
   Is 2001:123:4567:0:92df:5ff3:615f:6019 local? False
 fd31:aeb1:48df:0:3b14:e643:83d8:7017/64
   Is 10.75.22.10 local? False
   Is 2001:123:4567:0:92df:5ff3:615f:6019 local? False
 2001:123:4567:0:90cc:a629:cf42:5877/64
   Is 10.75.22.10 local? False
   Is 2001:123:4567:0:92df:5ff3:615f:6019 local? True
 fd0c:685d:4db3:4ba7:fb52:411:1c3c:8e9b/64
   Is 10.75.22.10 local? False
   Is 2001:123:4567:0:92df:5ff3:615f:6019 local? False
 fe80::a2d4:ede6:5587:6a7a%9/64
   Is 10.75.22.10 local? False
   Is 2001:123:4567:0:92df:5ff3:615f:6019 local? False

Name: enp2s0.1
Description: enp2s0.1
Addresses: 
```

I'd try to put together a PR to do all of this but I suspect building Sonarr is probably quite a bit bigger task to get going than my simple project above and I'd get up to my armpits in the weeds of that.

**Comment by thezoggy** [comment]
per irc, user just wants his statically assigned block from his isp to be viewed as local if sonarr sees traffic from within that prefix. 
i pushed for him to understand the security risk for those that dont have static ips nor the fact its not non routable space... ideally user should just use reverse proxy and set sonarr auth to external. and manage things that way.

**Comment by brianjmurrell** [comment]
> per irc, user just wants his statically assigned block from his isp to be viewed as local if sonarr sees traffic from within that prefix.

I just want the further the usability of Sonarr so that concept of _local network_ is correct such that it works for *everyone* now and forever, and is not just a guess/proxy for _local network_ that only includes people who are using non-routable and/or (only) link-local networks.

I fully understand the concept and motivation for Sonarr to enforce authentication when access is from a foreign network, but as implemented it only works for a(n admittedly large -- for now) subset of people.  But that subset is going to continually shrink as more and more ISPs, and therefore users, are assigned IPv6 prefixes for their LANs and such people stop using non-routable IPv4 addresses (that need NATting to the Internet).

If my proposed solution above were used instead of the non-routable-network test it would work for everyone (including everyone that the non-routable-network test is working for) now and forever.

> i pushed for him to understand the security risk for those that dont have static ips nor the fact its not non routable space...

But it's not a security risk.  Using ip-address/netmask matching is in fact a more secure and thorough test of whether a client is truly local to the Sonarr server's network or not.  Which use case(s) do you think it doesn't cover and is a security risk for?

**Comment by markus101** [comment]
> This is completely un-true. Even if the Sonarr machine were on a VPS with a publicly routable IP address, any client that is not on the same subnet (i.e. that were somewhere else on the Internet) would not match any of the ip-address/netmask pairs on the Sonarr machine and would therefore be subject to authentication.

Yeah, I misspoke, not everyone on the internet would have access, but everyone in the same subnet would have full, unauthenticated access.

> Which use case(s) do you think it doesn't cover and is a security risk for?

If my server has an IP of `198.1.2.4` in a `/24` network from a hosting provider that has other customers and someone else has their own server with the IP address of `198.1.2.3` which would also be a `/24` network, if Sonarr allowed all IPs within the same network unauthenticated access, because they're considered "local" to my server, my server would now allow full unauthenicated access from `192.1.2.3` and any of the other servers within that IP range, which would be 100% not what I want or expect.

While this detection would work for you because of your setup, it's not something we're going to implement because it would be far too easy for someone with a direct publicly accessible server (no NAT) to expose themselves unknowingly to others on the same network.

> Using ip-address/netmask matching is in fact a more secure and thorough test of whether a client is truly local to the Sonarr server's network or not.

It is, but the critical missing piece is knowing whether that network has foreign users on it, which cannot be known by this detection alone. Converting Sonarr to check that it's a private IP + on the same network would be more secure, but would drop support for cross-network routing (different subnets) or via VPN.

**Comment by bakerboy448** [comment]
Ref https://github.com/Radarr/Radarr/issues/9242

**Comment by markus101** [comment]
My example was with a server with a public IP in a class C network, where your IP "neighbours" would have access. ISPs/hosting providers aren't segmenting that IP space into small segments or server/router only networks like a VPN or guest AP would do. By validating whether an IP is in the same subnet in that scenario and allowing it unauthenticated access, not too far from what we saw with the rise of Cable internet and people with their unsecured computers directly connected sharing files to everyone on the network.

Allowing all RFC-1918 addresses is definitely not the most secure solution, also ensuring that it's part of the same network would definitely enhance that, but that'd two pieces and could break things in a VLAN/VPN scenario.

We've discussed this a bit further and at this time we're not going to make any changes at this time, though we may revisit how bypassing auth is handled in the future.
